### ISSUE TITLE
`have_enqueued_mail` raising undefined method `to_sym` for Hash

### ISSUE THREAD

**Issue by jasonkarns** [rspec/rspec-rails#2575]
## What Ruby, Rails and RSpec versions are you using?

```
ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-darwin20]
Rails 6.1.4.6
RSpec 3.10
  - rspec-core 3.10.2
  - rspec-expectations 3.10.2
  - rspec-mocks 3.10.3
  - rspec-rails 5.0.3
  - rspec-support 3.10.3
```

Issue does _not_ manifest on rspec-rails 5.0.2

## Observed behaviour

With the following spec:
```rb
       expect { subject.send_coach_lead_notification }
         .to have_enqueued_mail(CoachingInquiryMailer, :coach_notification_bookings_promo)
```

We get:

```
     NoMethodError:
       undefined method `to_sym' for {}:Hash
       Did you mean?  to_s
                      to_set
```

<details>
<summary>Full backtrace</summary>

<pre>
Run options: include {:locations=>{"./spec/services/coaching_inquiry_notifier_spec.rb"=>[111]}}

Randomized with seed 31320

CoachingInquiryNotifier
  when lead is from coach without offerings
    sends mailer with bookings promo (FAILED - 1)

Failures:

  1) CoachingInquiryNotifier when lead is from coach without offerings sends mailer with bookings promo
     Failure/Error:
       expect { subject.send_coach_lead_notification }
         .to have_enqueued_mail(CoachingInquiryMailer, :coach_notification_bookings_promo)

     NoMethodError:
       undefined method `to_sym' for {}:Hash
       Did you mean?  to_s
                      to_set
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/have_enqueued_mail.rb:148:in `block in deserialize_arguments'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/have_enqueued_mail.rb:148:in `each'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/have_enqueued_mail.rb:148:in `each_with_object'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/have_enqueued_mail.rb:148:in `deserialize_arguments'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/active_job.rb:148:in `arguments_match?'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/have_enqueued_mail.rb:89:in `arguments_match?'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/active_job.rb:104:in `block in check'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/active_job.rb:103:in `each'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/active_job.rb:103:in `partition'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/active_job.rb:103:in `check'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/active_job.rb:237:in `matches?'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/matchers/have_enqueued_mail.rb:40:in `matches?'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.2/lib/rspec/expectations/handler.rb:51:in `block in handle_matcher'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.2/lib/rspec/expectations/handler.rb:27:in `with_matcher'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.2/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.2/lib/rspec/expectations/expectation_target.rb:65:in `to'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.2/lib/rspec/expectations/expectation_target.rb:101:in `to'
     # ./spec/services/coaching_inquiry_notifier_spec.rb:112:in `block (3 levels) in <top (required)>'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:262:in `instance_exec'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:262:in `block in run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:508:in `block in with_around_and_singleton_context_hooks'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:465:in `block in with_around_example_hooks'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/hooks.rb:486:in `block in run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/hooks.rb:626:in `block in run_around_example_hooks_for'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:350:in `call'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-rails-5.0.3/lib/rspec/rails/adapters.rb:75:in `block (2 levels) in <module:MinitestLifecycleAdapter>'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:455:in `instance_exec'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:455:in `instance_exec'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/hooks.rb:390:in `execute_with'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/hooks.rb:628:in `block (2 levels) in run_around_example_hooks_for'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:350:in `call'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/hooks.rb:629:in `run_around_example_hooks_for'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/hooks.rb:486:in `run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:465:in `with_around_example_hooks'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:508:in `with_around_and_singleton_context_hooks'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example.rb:259:in `run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example_group.rb:644:in `block in run_examples'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example_group.rb:640:in `map'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example_group.rb:640:in `run_examples'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example_group.rb:606:in `run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example_group.rb:607:in `block in run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example_group.rb:607:in `map'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/example_group.rb:607:in `run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/runner.rb:121:in `block (3 levels) in run_specs'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/runner.rb:121:in `map'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/runner.rb:121:in `block (2 levels) in run_specs'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/configuration.rb:2067:in `with_suite_hooks'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/runner.rb:116:in `block in run_specs'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/reporter.rb:74:in `report'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/runner.rb:115:in `run_specs'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/runner.rb:89:in `run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/runner.rb:71:in `run'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/lib/rspec/core/runner.rb:45:in `invoke'
     # /usr/local/var/rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/rspec-core-3.10.2/exe/rspec:4:in `<top (required)>'
     # ./bin/rspec:29:in `load'
     # ./bin/rspec:29:in `<main>'

Finished in 0.72436 seconds (files took 2.59 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/services/coaching_inquiry_notifier_spec.rb:111 # CoachingInquiryNotifier when lead is from coach without offerings sends mailer with bookings promo

Randomized with seed 31320

</pre>
</details>

## Expected behaviour

This was a passing test before upgrading to rspec-rails 5.0.3

## Can you provide an example app?

https://github.com/jasonkarns/fluffy-octo-memory

I presume this is the result of #2566 (and related to #2570 ).

**Comment by pirj** [comment]
Thanks for reporting.
I wonder how this very basic case fails. It is pretty much covered by [this feature test](https://github.com/rspec/rspec-rails/blob/main/features/matchers/have_enqueued_mail_matcher.feature), and we run Ruby 2.7.5p203 with Rails 6.1.4.6, e.g. [fresh passing build](https://github.com/rspec/rspec-rails/runs/5279239365?check_suite_focus=true).

It would be very helpful if you could provide a reproducible example.

**Comment by jasonkarns** [comment]
@pirj pushed an rspec rails repro repo (say that again!) https://github.com/jasonkarns/fluffy-octo-memory

single model without anything special; single mailer. `rspec` from repo root will reproduce

**Comment by jasonkarns** [comment]
I just installed ruby 2.7.5 and still reproduced. (So you shouldn't need to downgrade to 2.7.3 if you don't wish to)

**Comment by pirj** [comment]
There's definitely a bug in this line:
```ruby
keywords.each_with_object({}) { |new_hash, keyword| puts new_hash, keyword; new_hash[keyword.to_sym] = hash[keyword] }
```
Block arguments should be the other way around, `|keyword, new_hash|`, and that fixes your case.

It's still surprising that this bug has gotten through all of our lines of defence, and how to write a spec to prevent this from re-appearing. Would you like to hack on it?

**Comment by JonRowe** [comment]
Its the result of a munged rubocop fix, it was initially a `reduce` which does have the args the other way round, so this was working but then broken to comply with the rubocop rule to not use `reduce` in these scenarios.

Spec wise we miss this because we don't test with an active record object, if you change the repro spec to use a string it doesn't encounter the bug.

I've tried to work up a repo with a fake global id job in our suite which isn't enough, and an active record one but it complains about it not being serialisable

**Comment by jasonkarns** [comment]
I was able to repro with a non-AR class by ensuring the job's param matched GlobalID::Identification:

https://github.com/jasonkarns/fluffy-octo-memory/commit/6dc4eabda20cd91160ca24221ad18f463adcf79c

```
class M
   include GlobalID::Identification
   def id
     1
   end
 end
```

`ReproMailer.with(inquiry: M.new).message.deliver_later`

**Comment by pirj** [comment]
Nice! Would you like to send a PR with an added spec for this?

**Comment by JonRowe** [comment]
I had a more complex example, but that actually works better, does not reproduce the issue within our test suite though

**Comment by jasonkarns** [comment]
Attempting to get local dev set up for rspec-rails to try a PR, but clone+bundle is failing :/ 

```
$ bundle install
Fetching https://github.com/rspec/rspec-support.git
Fetching https://github.com/rspec/rspec-mocks.git
Fetching https://github.com/rspec/rspec.git
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Bundler could not find compatible versions for gem "rspec-expectations":
  In Gemfile:
    rspec-expectations

    rspec-rails was resolved to 6.0.0.pre, which depends on
      rspec-expectations (= 3.12.0.pre)

Could not find gem 'rspec-expectations (= 3.12.0.pre)', which is required by gem 'rspec-rails', in source at `/Users/jasonkarns/Projects/rspec/rspec-expectations`.

The source contains the following gems matching 'rspec-expectations':
  * rspec-expectations-3.5.0.pre
```

I don't want to distract from this thread but if this is a known or quickfix...

**Comment by pirj** [comment]
Quick suggestion. Would 
       
    cd /Users/jasonkarns/Projects/rspec/rspec-expectations
    git checkout main
    git pull

work? It seems to find the repo locally, but the code there seems to be aged.

**Comment by jasonkarns** [comment]
Yep, apologies. I deleted my comment as soon as I realized it was due to old sibling repos.

**Comment by JonRowe** [comment]
The key to reproducing this was the fact it needed to be tested using a unified mailer which is a later Rails default than our base level

**Comment by JonRowe** [comment]
Closed in #2578 and released as 5.1.1
